File: zig/tests/clone_test.zig
const std = @import("std");

const clone = @import("gitologist").clone;

test "should clone a repository to default directory" {
    const io = std.testing.io;
    const allocator = std.testing.allocator;

    const tmp_path = try std.fs.path.join(allocator, &[_][]const u8{ "/tmp", "gitologist-clone-test-1" });
    defer allocator.free(tmp_path);

    const cwd = std.Io.Dir.cwd();

    try cwd.createDirPath(io, tmp_path);
    defer cwd.deleteTree(io, tmp_path) catch {};

    const target_path = try std.fs.path.join(allocator, &[_][]const u8{ tmp_path, "repo" });
    defer allocator.free(target_path);

    const result_path = try clone(io, allocator, "https://github.com/user/repo.git", target_path);
    defer allocator.free(result_path);

    try std.testing.expectEqualStrings(target_path, result_path);

    const git_dir_path = try std.fs.path.join(allocator, &[_][]const u8{ result_path, ".git" });
    defer allocator.free(git_dir_path);

    const git_dir = cwd.openDir(io, git_dir_path, .{}) catch {
        try std.testing.expect(false);
        return;
    };
    git_dir.close(io);
}

test "should clone to specified directory" {
    const io = std.testing.io;
    const allocator = std.testing.allocator;

    const tmp_path = try std.fs.path.join(allocator, &[_][]const u8{ "/tmp", "gitologist-clone-test-2" });
    defer allocator.free(tmp_path);

    const cwd = std.Io.Dir.cwd();

    try cwd.createDirPath(io, tmp_path);
    defer cwd.deleteTree(io, tmp_path) catch {};

    const target_path = try std.fs.path.join(allocator, &[_][]const u8{ tmp_path, "my-custom-dir" });
    defer allocator.free(target_path);

    const result_path = try clone(io, allocator, "https://github.com/user/repo.git", target_path);
    defer allocator.free(result_path);

    try std.testing.expectEqualStrings(target_path, result_path);

    _ = cwd.openDir(io, target_path, .{}) catch {
        try std.testing.expect(false);
        return;
    };
}

test "should extract repo name from URL" {
    const io = std.testing.io;
    const allocator = std.testing.allocator;

    const tmp_path = try std.fs.path.join(allocator, &[_][]const u8{ "/tmp", "gitologist-clone-test-3" });
    defer allocator.free(tmp_path);

    const cwd = std.Io.Dir.cwd();

    try cwd.createDirPath(io, tmp_path);
    defer cwd.deleteTree(io, tmp_path) catch {};

    const target_path = try std.fs.path.join(allocator, &[_][]const u8{ tmp_path, "test-repo" });
    defer allocator.free(target_path);

    const result_path = try clone(io, allocator, "https://github.com/user/my-repo.git", target_path);
    defer allocator.free(result_path);

    try std.testing.expectEqualStrings(target_path, result_path);
}

test "should initialize git repository" {
    const io = std.testing.io;
    const allocator = std.testing.allocator;

    const tmp_path = try std.fs.path.join(allocator, &[_][]const u8{ "/tmp", "gitologist-clone-test-4" });
    defer allocator.free(tmp_path);

    const cwd = std.Io.Dir.cwd();

    try cwd.createDirPath(io, tmp_path);
    defer cwd.deleteTree(io, tmp_path) catch {};

    const target_path = try std.fs.path.join(allocator, &[_][]const u8{ tmp_path, "test-repo" });
    defer allocator.free(target_path);

    const result_path = try clone(io, allocator, "https://github.com/user/repo.git", target_path);
    defer allocator.free(result_path);

    const git_dir_path = try std.fs.path.join(allocator, &[_][]const u8{ result_path, ".git" });
    defer allocator.free(git_dir_path);

    const git_dir = try cwd.openDir(io, git_dir_path, .{});
    defer git_dir.close(io);

    const head = try git_dir.readFileAlloc(io, "HEAD", allocator, .unlimited);
    defer allocator.free(head);

    try std.testing.expect(std.mem.indexOf(u8, head, "ref: refs/heads/master") != null);
}

test "should add remote" {
    const io = std.testing.io;
    const allocator = std.testing.allocator;

    const tmp_path = try std.fs.path.join(allocator, &[_][]const u8{ "/tmp", "gitologist-clone-test-5" });
    defer allocator.free(tmp_path);

    const cwd = std.Io.Dir.cwd();

    try cwd.createDirPath(io, tmp_path);
    defer cwd.deleteTree(io, tmp_path) catch {};

    const target_path = try std.fs.path.join(allocator, &[_][]const u8{ tmp_path, "test-repo" });
    defer allocator.free(target_path);

    const result_path = try clone(io, allocator, "https://github.com/user/repo.git", target_path);
    defer allocator.free(result_path);

    const git_dir_path = try std.fs.path.join(allocator, &[_][]const u8{ result_path, ".git" });
    defer allocator.free(git_dir_path);

    const git_dir = try cwd.openDir(io, git_dir_path, .{});
    defer git_dir.close(io);

    const config = try git_dir.readFileAlloc(io, "config", allocator, .unlimited);
    defer allocator.free(config);

    try std.testing.expect(std.mem.indexOf(u8, config, "[remote \"origin\"]") != null);
    try std.testing.expect(std.mem.indexOf(u8, config, "url = https://github.com/user/repo.git") != null);
}

test "should handle URLs with .git extension" {
    const io = std.testing.io;
    const allocator = std.testing.allocator;

    const tmp_path = try std.fs.path.join(allocator, &[_][]const u8{ "/tmp", "gitologist-clone-test-6" });
    defer allocator.free(tmp_path);

    const cwd = std.Io.Dir.cwd();

    try cwd.createDirPath(io, tmp_path);
    defer cwd.deleteTree(io, tmp_path) catch {};

    const target_path = try std.fs.path.join(allocator, &[_][]const u8{ tmp_path, "test-repo" });
    defer allocator.free(target_path);

    const result_path = try clone(io, allocator, "https://github.com/user/repo.git", target_path);
    defer allocator.free(result_path);

    try std.testing.expectEqualStrings(target_path, result_path);

    const git_dir_path = try std.fs.path.join(allocator, &[_][]const u8{ result_path, ".git" });
    defer allocator.free(git_dir_path);

    const git_dir = try cwd.openDir(io, git_dir_path, .{});
    defer git_dir.close(io);

    const config = try git_dir.readFileAlloc(io, "config", allocator, .unlimited);
    defer allocator.free(config);

    try std.testing.expect(std.mem.indexOf(u8, config, "url = https://github.com/user/repo.git") != null);
}

test "should handle URLs without .git extension" {
    const io = std.testing.io;
    const allocator = std.testing.allocator;

    const tmp_path = try std.fs.path.join(allocator, &[_][]const u8{ "/tmp", "gitologist-clone-test-7" });
    defer allocator.free(tmp_path);

    const cwd = std.Io.Dir.cwd();

    try cwd.createDirPath(io, tmp_path);
    defer cwd.deleteTree(io, tmp_path) catch {};

    const target_path = try std.fs.path.join(allocator, &[_][]const u8{ tmp_path, "test-repo" });
    defer allocator.free(target_path);

    const result_path = try clone(io, allocator, "https://github.com/user/repo", target_path);
    defer allocator.free(result_path);

    try std.testing.expectEqualStrings(target_path, result_path);

    const git_dir_path = try std.fs.path.join(allocator, &[_][]const u8{ result_path, ".git" });
    defer allocator.free(git_dir_path);

    const git_dir = try cwd.openDir(io, git_dir_path, .{});
    defer git_dir.close(io);

    const config = try git_dir.readFileAlloc(io, "config", allocator, .unlimited);
    defer allocator.free(config);

    try std.testing.expect(std.mem.indexOf(u8, config, "url = https://github.com/user/repo") != null);
}

test "should extract repo name from complex URL" {
    const io = std.testing.io;
    const allocator = std.testing.allocator;

    const tmp_path = try std.fs.path.join(allocator, &[_][]const u8{ "/tmp", "gitologist-clone-test-8" });
    defer allocator.free(tmp_path);

    const cwd = std.Io.Dir.cwd();

    try cwd.createDirPath(io, tmp_path);
    defer cwd.deleteTree(io, tmp_path) catch {};

    const target_path = try std.fs.path.join(allocator, &[_][]const u8{ tmp_path, "test-repo" });
    defer allocator.free(target_path);

    const result_path = try clone(io, allocator, "https://github.com/org/team/project.git", target_path);
    defer allocator.free(result_path);

    try std.testing.expectEqualStrings(target_path, result_path);
}

test "should handle subdirectory in URL" {
    const io = std.testing.io;
    const allocator = std.testing.allocator;

    const tmp_path = try std.fs.path.join(allocator, &[_][]const u8{ "/tmp", "gitologist-clone-test-9" });
    defer allocator.free(tmp_path);

    const cwd = std.Io.Dir.cwd();

    try cwd.createDirPath(io, tmp_path);
    defer cwd.deleteTree(io, tmp_path) catch {};

    const target_path = try std.fs.path.join(allocator, &[_][]const u8{ tmp_path, "test-repo" });
    defer allocator.free(target_path);

    const result_path = try clone(io, allocator, "https://github.com/user/nested/project.git", target_path);
    defer allocator.free(result_path);

    try std.testing.expectEqualStrings(target_path, result_path);
}

test "should throw error if directory already exists" {
    const io = std.testing.io;
    const allocator = std.testing.allocator;

    const tmp_path = try std.fs.path.join(allocator, &[_][]const u8{ "/tmp", "gitologist-clone-test-10" });
    defer allocator.free(tmp_path);

    const cwd = std.Io.Dir.cwd();

    try cwd.createDirPath(io, tmp_path);
    defer cwd.deleteTree(io, tmp_path) catch {};

    const existing_path = try std.fs.path.join(allocator, &[_][]const u8{ tmp_path, "repo" });
    defer allocator.free(existing_path);

    try cwd.createDirPath(io, existing_path);

    const result = clone(io, allocator, "https://github.com/user/repo.git", existing_path);
    try std.testing.expectError(error.DestinationPathAlreadyExists, result);
}

File: web/tests/clone.test.ts
import { mkdir, rm, readFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { describe, it, expect, beforeEach, afterEach } from "vite-plus/test";

import { clone } from "../src/clone";

describe("clone", () => {
	let testDir: string;

	beforeEach(async () => {
		testDir = join(
			tmpdir(),
			`gitologist-test-${Date.now()}-${Math.random().toString(36).slice(2)}`,
		);
		await mkdir(testDir, { recursive: true });
	});

	afterEach(async () => {
		await rm(testDir, { recursive: true, force: true });
	});

	it("should clone a repository to default directory", async () => {
		const url = "https://github.com/user/repo.git";
		const targetPath = join(testDir, "repo");
		const resultPath = await clone(url, targetPath);

		expect(resultPath).toBe(targetPath);

		const gitDir = join(resultPath, ".git");
		const { default: fs } = await import("node:fs");
		expect(fs.existsSync(gitDir)).toBe(true);
	});

	it("should clone to specified directory", async () => {
		const url = "https://github.com/user/repo.git";
		const targetPath = join(testDir, "my-custom-dir");
		const resultPath = await clone(url, targetPath);

		expect(resultPath).toBe(targetPath);

		const { default: fs } = await import("node:fs");
		expect(fs.existsSync(targetPath)).toBe(true);
	});

	it("should extract repo name from URL", async () => {
		const url = "https://github.com/user/my-repo.git";
		const targetPath = join(testDir, "test-repo");
		const resultPath = await clone(url, targetPath);

		expect(resultPath).toBe(targetPath);
	});

	it("should initialize git repository", async () => {
		const url = "https://github.com/user/repo.git";
		const targetPath = join(testDir, "test-repo");
		const resultPath = await clone(url, targetPath);

		const headPath = join(resultPath, ".git", "HEAD");
		const headContent = await readFile(headPath, "utf-8");

		expect(headContent).toContain("ref: refs/heads/main");
	});

	it("should add remote", async () => {
		const url = "https://github.com/user/repo.git";
		const targetPath = join(testDir, "test-repo");
		const resultPath = await clone(url, targetPath);

		const configPath = join(resultPath, ".git", "config");
		const configContent = await readFile(configPath, "utf-8");

		expect(configContent).toContain('[remote "origin"]');
		expect(configContent).toContain("url = https://github.com/user/repo.git");
	});

	it("should handle URLs with .git extension", async () => {
		const url = "https://github.com/user/repo.git";
		const targetPath = join(testDir, "test-repo");
		const resultPath = await clone(url, targetPath);

		expect(resultPath).toBe(targetPath);

		const configPath = join(resultPath, ".git", "config");
		const configContent = await readFile(configPath, "utf-8");

		expect(configContent).toContain("url = https://github.com/user/repo.git");
	});

	it("should handle URLs without .git extension", async () => {
		const url = "https://github.com/user/repo";
		const targetPath = join(testDir, "test-repo");
		const resultPath = await clone(url, targetPath);

		expect(resultPath).toBe(targetPath);

		const configPath = join(resultPath, ".git", "config");
		const configContent = await readFile(configPath, "utf-8");

		expect(configContent).toContain("url = https://github.com/user/repo");
	});

	it("should extract repo name from complex URL", async () => {
		const url = "https://github.com/org/team/project.git";
		const targetPath = join(testDir, "test-repo");
		const resultPath = await clone(url, targetPath);

		expect(resultPath).toBe(targetPath);
	});

	it("should handle subdirectory in URL", async () => {
		const url = "https://github.com/user/nested/project.git";
		const targetPath = join(testDir, "test-repo");
		const resultPath = await clone(url, targetPath);

		expect(resultPath).toBe(targetPath);
	});

	it("should throw error if directory already exists", async () => {
		const url = "https://github.com/user/repo.git";
		const existingPath = join(testDir, "repo");
		await mkdir(existingPath, { recursive: true });

		await expect(clone(url, existingPath)).rejects.toThrow();
	});
});
